### PR TITLE
Interpolation image

### PR DESCRIPTION
**Pull request by lianabagh**

### CODE REVIEW

File: yield_forecasting/utils/image_interpolation/interpolation.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        x_input, y_input = np.where((cropped_boundary > 0) & (~np.isnan(cropped_img)))
+        predicted_values = make_interpolated_image(cropped_img, x_input, y_input, x_new_predicted, y_new_predicted)
+        cropped_img[x_new_predicted, y_new_predicted] = predicted_values
+        cropped_img[np.isnan(cropped_img)] = predicted_values[~np.isnan(predicted_values)].mean()
+        interpolated_img[strided_bbox] = cropped_img
+
+    return interpolated_img
+
+
+if __name__ == '__main__':
```
Comment (razmikmelikbekyan) [inline yield_forecasting/utils/image_interpolation/interpolation.py:91]: please remove script part from here

File: yield_forecasting/utils/image_interpolation/interpolation.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    start_index_y = (bbox[0] - IMAGE_PADDING_SIZE) if (bbox[0] - IMAGE_PADDING_SIZE) >= 0 else 0
+    end_index_x = (bbox[1] + bbox[3] + IMAGE_PADDING_SIZE) if (bbox[1] + bbox[3] + IMAGE_PADDING_SIZE) < image_shape[
+        0] else image_shape[0]
+    end_index_y = (bbox[0] + bbox[2] + IMAGE_PADDING_SIZE) if (bbox[0] + bbox[2] + IMAGE_PADDING_SIZE) < image_shape[
+        1] else image_shape[1]
+    index_coordinates = np.s_[start_index_x:end_index_x, start_index_y:end_index_y]
+    return index_coordinates
+
+
+def get_interpolated_image(img: np.ndarray, b_mask: np.ndarray, no_data_value: float):
```
Comment (razmikmelikbekyan) [inline yield_forecasting/utils/image_interpolation/interpolation.py:65]: please make an additional function with input like this: image_path: str, boundary_mask_path: str, output_path: str

File: yield_forecasting/utils/image_interpolation/interpolation.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    Returns:
+        predicted values after interpolation
+    """
+    samples = img[x_index_input, y_index_input]
+    predicted_values = griddata((x_index_input, y_index_input), samples, (x_index_predict, y_index_predict))
+    return predicted_values
+
+
+def get_padded_coordinates(bbox, image_shape):
```
Comment (razmikmelikbekyan) [inline yield_forecasting/utils/image_interpolation/interpolation.py:45]: typing

File: yield_forecasting/utils/image_interpolation/interpolation.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    """
+
```
Comment (razmikmelikbekyan) [inline yield_forecasting/utils/image_interpolation/interpolation.py:47]: description

File: yield_forecasting/utils/image_interpolation/interpolation.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    Returns:
+
+    """
+    with rasterio.open(path) as src_dataset:
+        img = np.squeeze(src_dataset.read())
+        profile = src_dataset.profile
+    return img, profile
+
+
+def make_interpolated_image(img: np.ndarray, x_index_input: np.ndarray, y_index_input: np.ndarray,
```
Comment (razmikmelikbekyan) [inline yield_forecasting/utils/image_interpolation/interpolation.py:27]: img -> image

File: yield_forecasting/utils/image_interpolation/interpolation.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+import cv2
+import numpy as np
+import rasterio
+from scipy.interpolate import griddata
+
+IMAGE_PADDING_SIZE = 20
+
+
+def read_image(path: str):
```
Comment (razmikmelikbekyan) [inline yield_forecasting/utils/image_interpolation/interpolation.py:12]: image_path

File: yield_forecasting/utils/image_interpolation/interpolation.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        0] else image_shape[0]
+    end_index_y = (bbox[0] + bbox[2] + IMAGE_PADDING_SIZE) if (bbox[0] + bbox[2] + IMAGE_PADDING_SIZE) < image_shape[
+        1] else image_shape[1]
+    index_coordinates = np.s_[start_index_x:end_index_x, start_index_y:end_index_y]
+    return index_coordinates
+
+
+def get_interpolated_image(img: np.ndarray, b_mask: np.ndarray, no_data_value: float):
+    """
+
```
Comment (razmikmelikbekyan) [inline yield_forecasting/utils/image_interpolation/interpolation.py:67]: please describe a bit a logic

File: yield_forecasting/utils/image_interpolation/interpolation.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
+import argparse
```
Comment (razmikmelikbekyan) [inline yield_forecasting/utils/image_interpolation/interpolation.py:1]: please just put inside `utils`, without additional folder crateion

File: tests/unit/image_interpolation/test_interpolation.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+from yield_forecasting.utils.image_interpolation.interpolation import make_interpolated_image
```
Comment (razmikmelikbekyan) [inline tests/unit/image_interpolation/test_interpolation.py:1]: test folder structure should have the same structure as the code, so it should be like this - test/unit/yield_forecasting/utils...

File: tests/unit/image_interpolation/test_interpolation.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+
+
+def test_padded_shapes():
```
Comment (razmikmelikbekyan) [inline tests/unit/image_interpolation/test_interpolation.py:4]: please add a test, that checks all the data is filed within boundary mask, feel free to use the images that I have provided

**Review by razmikmelikbekyan — CHANGES_REQUESTED** [review]

**Review by razmikmelikbekyan — APPROVED** [review]
